src/main.cc

```diff
@@ -911,6 +936,31 @@ mainReconfigureStart(void)
              false);
 }
 
+// TODO: move here the rest of the common startup/reconfiguration routines
+/// configures and starts services common for both startup and reconfiguration
+static void
+startServices()
+{
+    if (IamPrimaryProcess()) {
+#if USE_WCCP
+        wccpInit();
+#endif
+#if USE_WCCPv2
+        wccp2Init();
+#endif
+    }
+
+    if (IamWorkerProcess())
+        startWorkerServices();
+
+    if (Store::Controller::WaitingForIndex()) {
+        debugs(1, DBG_IMPORTANT, "Waiting for Store indexing completion before opening listening sockets "
+               "and/or contacting cache_peers");
+        return;
+    }
+    RunRegisteredHere(RegisteredRunner::useFullyIndexedStore);
```

Comment:
>  we cannot tell that the store is fully indexed just because Store::Controller::WaitingForIndex() is false.

It is false when either the indexing is already completed or no indexing is expected (no dirs were configured). The first case is processed in storeCleanup() which is scheduled by an event. The second case is processed here.

If I understood your concern correctly, there may be a situation when we call useFullyIndexedStore() twice - i.e., we call it here (the second time) when the indexing is expected and has been already completed. However, this should not be possible because the scheduled event (storeCleanup) will be activated only _after_ all this synchronous initializing code is completed.

On the other hand, I am not against refactoring it to make the code safer (e.g., it is not prohibited to call storeCleanup() synchronously with the initialization code someday).

Reply:
```C++
if (!Store::Controller::WaitingForIndex()) // tests !A
    RunRegisteredHere(useFullyIndexedStore); // makes B (i.e. "Store is fully indexed") true
```

> If I understood your concern correctly, there may be a situation when we call useFullyIndexedStore() twice

My concern here is that the code looks illogical. Whether the code happens to "work" or, if it does not work, what exact bugs those logic violations lead to, I do _not_ know -- I cannot track all the low-level dependencies. Hopefully, there is a solution that makes such tracking unnecessary for the code to "work" and make sense to a reasonable reader.

```C++
/// whether we are currently building the Store index in foreground mode
static bool WaitingForIndex() { return opt_foreground_rebuild && store_dirs_rebuilding; }
```

> Store::Controller::WaitingForIndex() is false when either the indexing is already completed or no indexing is expected (no dirs were configured).

FWIW, the above statement does not seem to match WaitingForIndex() description and implementation. For example, WaitingForIndex() is false when opt_foreground_rebuild is not set, regardless of the current Store index rebuilding state or necessity. I am not sure, but I think the WaitingForIndex() description and implementation are correct.

Reply:
> The "nothing to index" case occurs when Config.cacheSwap.swapDirs is null i.e., there were no configured cache dirs at all. In this case, we should not get into the loop over cache dirs in Disks::init().

I am worried about the case where there are configured cache_dirs, but there is still nothing to index because all those configured cache_dirs are inactive. There is such a thing as "indexing state", and some of our changes revolve around that state. It feels like there are only two reasonable choices with regard to creating that state, but we are using none of these:

* Never index nothing: When there are no cache_dirs to index, the process should not enter the indexing state at all. This approach reduces noise by eliminating a largely useless state.

* Always index: The process should always enter (and then later exit) the indexing state. When there are no cache_dirs to index, that state would be very short, of course. This approach avoids creating a special case that depends on the presence of configured cache_dirs and the role of the process. Naturally, special cases are easy to mishandle (as your fix in branch commit 6a69605 illustrates).

I suspect there is a way to adjust store_dirs_rebuilding maintenance to avoid indexing nothing and confine the special "nothing to index" case to one place in internal Store code where it would be easy to handle correctly.

The correct solution may depend on what happens during _reconfiguration_ when new cache_dirs are added: Does Squid call Store::Disks::init() in that case? If not, what code calls Disk::init() to initialize the newly added cache_dir? Can reconfiguraion trigger a useFullyIndexedStore event?

Reply:
> during reconfiguration when new cache_dirs are added

No new dirs are added during reconfiguration: neither Store::Disks::init() is called in this case, no individual Disk::init().

Reply:
> No new dirs are added during reconfiguration

What prevents their addition at the bottom of Store::Disks::Parse()?

> neither Store::Disks::init() is called in this case, no individual Disk::init().

That may be true. Squid may be adding new cache_dirs during reconfiguration -- calling createSwapDir() and adding the pointers to swap.swapDirs -- but not initializing the newly created disks properly. I doubt we should address this problem in this PR, but, if we can, we should mark it so that we can rely on a particular future solution of this problem.

I suggest adding the following comment to the bottom of Store::Disks::Parse():

```C++
    // XXX: cache_dirs added during reconfiguration are not Disk::init()ialized.
    // TODO: Prohibit such a heavy (and already broken) operation during reconfiguration.
    // create a new cache_dir
    allocate_new_swapdir(swap);
```